google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java

```diff
@@ -459,39 +463,55 @@ private static boolean isUnaryComparison(@Nullable Object value) {
     return value == null || value.equals(Double.NaN) || value.equals(Float.NaN);
   }
 
-  /** Computes the backend ordering semantics for DocumentSnapshot cursors. */
-  private ImmutableList<FieldOrder> createImplicitOrderBy() {
-    List<FieldOrder> implicitOrders = new ArrayList<>(options.getFieldOrders());
-
-    // If no explicit ordering is specified, use the first inequality to define an implicit order.
-    if (implicitOrders.isEmpty()) {
-      for (FilterInternal filter : options.getFilters()) {
-        FieldReference fieldReference = filter.getFirstInequalityField();
-        if (fieldReference != null) {
-          implicitOrders.add(new FieldOrder(fieldReference, Direction.ASCENDING));
-          break;
+  /** Returns the sorted set of inequality filter fields used in this query. */
+  private SortedSet<FieldPath> getInequalityFilterFields() {
+    SortedSet<FieldPath> result = new TreeSet<>();
+
+    for (FilterInternal filter : options.getFilters()) {
+      for (FieldFilterInternal subFilter : filter.getFlattenedFilters()) {
+        if (subFilter.isInequalityFilter()) {
+          result.add(FieldPath.fromServerFormat(subFilter.fieldReference.getFieldPath()));
         }
       }
     }
 
-    boolean hasDocumentId = false;
-    for (FieldOrder fieldOrder : implicitOrders) {
-      if (FieldPath.isDocumentId(fieldOrder.fieldReference.getFieldPath())) {
-        hasDocumentId = true;
+    return result;
+  }
+
+  /** Computes the backend ordering semantics for DocumentSnapshot cursors. */
+  private ImmutableList<FieldOrder> createImplicitOrderBy() {
+    // Any explicit order by fields should be added as is.
+    List<FieldOrder> result = new ArrayList<>(options.getFieldOrders());
+
+    HashSet<String> fieldsNormalized = new HashSet<>();
+    for (FieldOrder order : result) {
+      fieldsNormalized.add(order.fieldReference.getFieldPath());
+    }
+
+    /** The order of the implicit ordering always matches the last explicit order by. */
+    Direction lastDirection =
+        result.isEmpty() ? Direction.ASCENDING : result.get(result.size() - 1).direction;
+
+    /**
+     * Any inequality fields not explicitly ordered should be implicitly ordered in a
+     * lexicographical order. When there are multiple inequality filters on the same field, the
+     * field should be added only once. Note: `SortedSet<FieldPath>` sorts the key field before
+     * other fields. However, we want the key field to be sorted last.
+     */
+    SortedSet<FieldPath> inequalityFields = getInequalityFilterFields();
+    for (FieldPath field : inequalityFields) {
+      if (!fieldsNormalized.contains(field.toString())
+          && !FieldPath.isDocumentId(field.toString())) {
```

Comment:
ultra-nit: on line 503 and 504 and 510, it's better to use `.canonicalString()` (similar to Android) rather than `.toString()`. 

Today these two functions do the same thing, but `toString()` is not unlikely to change.

Reply:
`.canonicalString` is a private method:  https://github.com/googleapis/java-firestore/blob/01f669caca34390eafb3a82f842a400f87759bd5/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldPath.java#L133

It is called inside `.getEncodedPath` with memoization. So simply turning `.canonicalString` to public would not be ideal neither.

Reply:
Ah. Didn't notice that. SG